discv5/discv5-wire.md

```diff
@@ -1,395 +1,307 @@
-not backward compatible.
-
-Another key difference from Discovery V4 is that here we begin to
-introduce a level of abstraction above UDP. Discovery V5 should not be
-restricted to UDP. In this specification *packet* and *message* are not
-necessarily synonymous.
-
-**TBD: Decide if to permit future streamed transports or not.**
-
-However, UDP poses some challenges when it comes to avoiding
-fragmentation and transmitting large messages of a lossy transport,
-which are considered next.
-
-### UDP transmission strategy
-
-So that the message types and their exchange patterns are not tied to a
-transport (eg: UDP), we should avoid introdcuing packet types that only
-deal with UDP fragmentation and reliability of larger packets. In other
-words, it would be best to avoid "ACK" type packets, as these become
-redundant on TCP or other future transports.
+Node Discovery Protocol v5 - Wire Protocol
+==========================================
 
-`FINDNODE` needs to be able to return up to `k` ENR records, plus other
-data, and `TOPICQUERY` in the form proposed in this document may also
-distribute a significant list of ENRs. As per specification the maximum
-size of an ENR is 300 bytes. Because `k` is typically 16 in Ethereum,
-a message will be at least 4800 bytes, not including additional data such
-as the header.
+**Draft of April 2019**
 
-This will cause packet fragmentation or packet loss as it exceeds the
-typical MTU of 1500 at the time of writing. We will assume a 'healthy' MTU of 1280,
-conforming with v4 expectations.
+This document specifies the wire protocol of Node Discovery v5. Note that this
+specification is a work in progress and may change incompatibly without prior notice.
 
-With the above rationale, the proposal for a transport agnostic way of
-responding with large messages is as follows:
+UDP Communication
+-----------------
 
-- Potentially large response messages (eg: `TOPICQUERY` and `FINDNODE`
-  responses) can be sent as *multiple messages*.
-- These are complete messages with their own header.
-- Individual messages in a multi-message response of length `t`, will
-  have a message counter `n`, and each response message will have a
-  message conveying the information that this one is `n` of `t`.
-- While `n` is not (at the time of writing) necessary information, it
-  may be a useful forward compatibility feature in the case that
-  ordering of the response stream is important.
-- The recipient of a multi-message response must expect that some
-  messages will be lost over UDP. Implementations must receive with a
-  timeout.
-- The *conversation nonce* determines the conversation and is used on
-  each message.
-- The precise same message format may be used over TCP, except that
-  the message may include a message "1 of 1" descriptor, include all
-  data in the response stream, and rely on length information
+Node discovery messages are sent as UDP datagrams. Since UDP is a lossy transport, packets
```

Comment:
I think we do not need to tie the protocol to UDP at this point. This could be rephrased to "Node discovery messages are sent over lossy transports like UDP. Packets may be received..."